### ISSUE TITLE
Is there earlystopping in yolov5?

### ISSUE THREAD

**Issue by siyangxie** [ultralytics/yolov5#1633]
## ❔Question
Trying to iterate over 300 epochs when only 100 could have worked does not make sense to me. I'm just wondering if there is earlystopping like what they have with tensorflow? 

Many thanks.

**Comment by yinyuhangi** [comment]
I think just stop training because it will save best.pt automatically.

**Comment by glenn-jocher** [comment]
@SiyangXie best.pt will always contain the best model, as @yinyuhangi no matter when your training stops (or crashes). Since we use a cosine LR scheduler, it's most effective to train for the number of epochs you actually intend to train for, as the LR gradually decreases towards zero.

**Comment by leeyunhome** [comment]
> @SiyangXie best.pt will always contain the best model, as @yinyuhangi no matter when your training stops (or crashes). Since we use a cosine LR scheduler, it's most effective to train for the number of epochs you actually intend to train for, as the LR gradually decreases towards zero.

Hello, @glenn-jocher 

Still, if you specify too few epochs, it seems that learning will not work.

Is there a recommended number of epochs?

Thank you.

**Comment by robotwhispering** [comment]
I noticed that if I stop my "from scratch" models early, the file size is often > 4x what what it is when the saves the file at the end of training. Could you just start another training for 2 epochs and get that same  trimming effect?

**Comment by glenn-jocher** [comment]
@robotwhispering 👋 Hello! Thanks for asking about **training checkpoints**.  YOLOv5 🚀 checkpoints should be about 4X the size of final trained checkpoints, as they carry not just a FP16 model, but a FP16 [EMA](https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage) and an FP32 optimizer of the same size as the model (each model parameter has its own FP32 gradient saved within the optimizer). Final checkpoints contain only an FP16 model, with the EMA and optimizer both stripped after the final epoch of training:
https://github.com/ultralytics/yolov5/blob/b4a29b5a8d63a8c2d4a8929942b44e8969c5dddd/train.py#L423-L425

The `strip_optimizer()` function updates the checkpoint dictionary value for `model`, replacing it with `ema`, and sets `ema` and `optimizer` keys to `None`, which will reduce the checkpoint size by 3/4.
https://github.com/ultralytics/yolov5/blob/b4a29b5a8d63a8c2d4a8929942b44e8969c5dddd/utils/general.py#L741-L755

Final trained checkpoints after `strip_optimizer()` should match the README table sizes below.
<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

You can also run `strip_optimizer()` manually on any checkpoint to convert it into finalized weights ready to train a new model:
```python
from utils.general import strip_optimizer

strip_optimizer('path/to/best.pt')
!python train.py --weights path/to/best.pt  # use best.pt to train a new model
```

### Model Exports

Exported models may be larger or smaller than their PyTorch starting points depending on their quantization. i.e. FP16 models will be the same size, FP32 models will be 2X the size, and INT8 models will be 0.5X the size. See YOLOv5 [Export tutorial](https://docs.ultralytics.com/yolov5/tutorials/model_export) for quantization options on model export.

Good luck 🍀 and let us know if you have any other questions!

**Comment by robotwhispering** [comment]
AWESOME, THANKS! Based on your suggestion, I found the strip optimizer definition and implementation,  and made a quick little script to go handle it anytime Colab kicks me off unexpectedly.
